README.md
# K8s service broadcasting

[![CircleCI](https://img.shields.io/circleci/build/github/FUSAKLA/k8s-service-broadcasting/master?style=flat-square)](https://app.circleci.com/github/FUSAKLA/k8s-service-broadcasting/pipelines?branch=master)
[![Docker Pulls](https://img.shields.io/docker/pulls/fusakla/k8s-service-broadcasting?style=flat-square)](https://hub.docker.com/repository/docker/fusakla/k8s-service-broadcasting)

Tool allowing to broadcast/mirror/duplicate HTTP requests to all ready endpoints of a Kubernetes service.

It supports two modes:
 - _Consistency_: Waits for response from all service endpoints and and response with one of erroneous response otherwise one of the succeeded. 
 - _Availability_: Responses with first successful response if any, otherwise with one of the error messages.

This behaviour can be controlled by the `--all-must-succeed` flag. Defaults to `true`.

## Usage

```bash
$ ./k8s-service-broadcasting --help
Tool allowing to broadcast/mirror/duplicate HTTP requests to all endpoints of Kubernetes service.
Waits for all of them to end and reports back failed request if any. If not returns last successful.

Usage:
  k8s-service-broadcasting [flags]

Flags:
      --all-must-succeed           By default if any backend fails, the whole request fails. If disabled one succeeded response is enough. (default true)
  -h, --help                       help for k8s-service-broadcasting
  -i, --interface string           Interface to listen on. (default "0.0.0.0:8080")
      --keepalive                  If keepalive should be enabled. (default true)
  -k, --kubeconfig string          Location of the kubeconfig, default if in cluster config or value of KUBECONFIG env variable. (default "/home/fusakla/.kube/conf/kubeconfig.yaml")
  -l, --log-level string           Log level (debug, info, warning, ...) default info. (default "info")
  -m, --metrics-interface string   Interface for exposing metrics. (default "0.0.0.0:8081")
  -n, --namespace string           Namespace to watch for.
  -p, --port-name string           Name of service port to sed the requests to.
  -s, --service string             Name of service to sed the requests to.
  -t, --timeout duration           Timeout for mirrored requests. (default 10s)
```

## Instrumentation
There is second interface for exposing Prometheus metrics and providing lifecycle endpoints.
This interface is running by default on `0.0.0.0:8081`. The endpointa are:
- `/metrics` Prometheus metrics
- `/-/healthy` liveness probe
- `/-/ready` readiness probe 

## Build
**single binary**
```bash
make build
```

## Deployment
See the [`kubernetes/`](./kubernetes) folder with example.


CHANGELOG.md

## Unreleased

## 0.1.0 / 2020-1-26

Initial release


pkg/handler_test/handler_test.go
// Copyright 2019 FUSAKLA Martin Chodúr
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package handler_test

import (
	"fmt"
	"github.com/fusakla/k8s-service-broadcasting/pkg/handler"
	"github.com/magiconair/properties/assert"
	log "github.com/sirupsen/logrus"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func init() {
	log.SetLevel(log.DebugLevel)
}

func getServerURL(URL string) string {
	return strings.TrimPrefix(URL, "http://")
}

type testCase struct {
	addresses      []string
	allMustSucceed bool
	keepalive      bool
	response       int
	timeout        time.Duration
}

func TestMultiplexingHandler_ServeHTTP(t *testing.T) {
	var (
		okServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			_, err := fmt.Fprintln(w, "OK", http.StatusOK)
			if err != nil {
				t.Error(err)
			}
		}))
		errServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			http.Error(w, "fail", http.StatusServiceUnavailable)
		}))
		okServerURL  = getServerURL(okServer.URL)
		errServerURL = getServerURL(errServer.URL)

		testCases = []testCase{

			{addresses: []string{okServerURL}, timeout: 0, allMustSucceed: true, keepalive: false, response: http.StatusGatewayTimeout},
			{addresses: []string{}, timeout: 60 * time.Second, allMustSucceed: true, keepalive: false, response: http.StatusServiceUnavailable},

			{addresses: []string{okServerURL}, timeout: 60 * time.Second, allMustSucceed: true, keepalive: false, response: http.StatusOK},
			{addresses: []string{okServerURL, okServerURL}, timeout: 60 * time.Second, allMustSucceed: true, keepalive: false, response: http.StatusOK},
			{addresses: []string{okServerURL, okServerURL}, timeout: 60 * time.Second, allMustSucceed: false, keepalive: false, response: http.StatusOK},
			{addresses: []string{okServerURL, okServerURL}, timeout: 60 * time.Second, allMustSucceed: true, keepalive: true, response: http.StatusOK},

			{addresses: []string{errServerURL}, timeout: 60 * time.Second, allMustSucceed: true, keepalive: false, response: http.StatusServiceUnavailable},
			{addresses: []string{errServerURL}, timeout: 60 * time.Second, allMustSucceed: false, keepalive: false, response: http.StatusServiceUnavailable},
			{addresses: []string{errServerURL, errServerURL}, timeout: 60 * time.Second, allMustSucceed: true, keepalive: true, response: http.StatusServiceUnavailable},

			{addresses: []string{okServerURL, errServerURL}, timeout: 60 * time.Second, allMustSucceed: true, keepalive: false, response: http.StatusServiceUnavailable},
			{addresses: []string{okServerURL, errServerURL}, timeout: 60 * time.Second, allMustSucceed: false, keepalive: false, response: http.StatusOK},
		}
	)
	defer okServer.Close()
	defer errServer.Close()

	fmt.Printf("okServer URL: %s\n", okServerURL)
	fmt.Printf("errServer URL: %s\n", errServerURL)

	for _, testCase := range testCases {
		fmt.Printf("\n\ntesting hosts: %v expected: %v\n", testCase.addresses, testCase.response)
		multiplexingHandler := handler.NewMultiplexingHandler("", testCase.timeout, testCase.allMustSucceed, testCase.keepalive)
		testedServer := httptest.NewServer(multiplexingHandler)
		multiplexingHandler.SetOwnAddress(testedServer.URL)
		multiplexingHandler.SetTargetAddresses(testCase.addresses)
		response, err := http.Get(testedServer.URL)
		if err != nil {
			t.Error(err)
		}
		testedServer.Close()
		assert.Equal(t, response.StatusCode, testCase.response)
		body, err := ioutil.ReadAll(response.Body)
		if err != nil {
			body = []byte{}
		}
		fmt.Printf("response status: %v body: %s\n", response.StatusCode, string(body))
	}

}


pkg/handler/util.go
// Copyright 2019 FUSAKLA Martin Chodúr
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package handler

import (
	"bytes"
	"fmt"
	log "github.com/sirupsen/logrus"
	"io"
	"io/ioutil"
	"math/rand"
	"net/http"
	"net/url"
)

func setRequestTarget(request *http.Request, target string, scheme string) error {
	URL, err := url.Parse(fmt.Sprintf("%s://%s%s", scheme, target, request.URL.String()))
	if err != nil {
		return err
	}
	request.URL = URL
	return nil
}

func duplicateRequest(request *http.Request) (dup *http.Request) {
	var bodyBytes []byte
	if request.Body != nil {
		bodyBytes, _ = ioutil.ReadAll(request.Body)
	}
	request.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
	dup = &http.Request{
		Method:        request.Method,
		URL:           request.URL,
		Proto:         request.Proto,
		ProtoMajor:    request.ProtoMajor,
		ProtoMinor:    request.ProtoMinor,
		Header:        request.Header,
		Body:          ioutil.NopCloser(bytes.NewBuffer(bodyBytes)),
		Host:          request.Host,
		ContentLength: request.ContentLength,
		Close:         true,
	}
	return
}

func newResponse(status int, message string) *http.Response {
	return &http.Response{
		StatusCode: status,
		Body:       ioutil.NopCloser(bytes.NewReader([]byte(message))),
		Request:    &http.Request{URL: &url.URL{}},
	}
}

func randomResponse(responses []*http.Response) *http.Response {
	if len(responses) == 0 {
		return nil
	}
	return responses[rand.Intn(len(responses))]
}

func sendResponse(w http.ResponseWriter, resp *http.Response) {
	defer resp.Body.Close()
	for k, v := range resp.Header {
		w.Header()[k] = v
	}
	w.WriteHeader(resp.StatusCode)
	if _, err := io.Copy(w, resp.Body); err != nil {
		log.Errorf("failed to write the response: %v", err)
	}
}


pkg/handler/handler.go
// Copyright 2019 FUSAKLA Martin Chodúr
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package handler

import (
	"bytes"
	"context"
	"fmt"
	"github.com/google/uuid"
	"github.com/prometheus/client_golang/prometheus"
	log "github.com/sirupsen/logrus"
	"io/ioutil"
	"math/rand"
	"net"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

var (
	requestDurationSeconds = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name: "request_duration_seconds",
			Help: "Duration of HTTP requests.",
		},
		[]string{"type", "endpoint", "status_code"},
	)
)

func init() {
	prometheus.MustRegister(requestDurationSeconds)
}

func NewMultiplexingHandler(ownAddress string, timeout time.Duration, allMustSucceed, keepalive bool) *multiplexingHandler {
	return &multiplexingHandler{
		ownAddress:           ownAddress,
		timeout:              timeout,
		allMustSucceed:       allMustSucceed,
		keepalive:            keepalive,
		targetAddresses:      &[]string{},
		targetAddressesMutex: sync.Mutex{},
	}
}

type multiplexingHandler struct {
	ownAddress           string
	timeout              time.Duration
	allMustSucceed       bool
	keepalive            bool
	targetAddresses      *[]string
	targetAddressesMutex sync.Mutex
}

func (h *multiplexingHandler) GetTargetAddresses() []string {
	h.targetAddressesMutex.Lock()
	defer h.targetAddressesMutex.Unlock()
	return *h.targetAddresses
}

func (h *multiplexingHandler) SetTargetAddresses(addresses []string) {
	h.targetAddressesMutex.Lock()
	defer h.targetAddressesMutex.Unlock()
	h.targetAddresses = &addresses
}

func (h *multiplexingHandler) SetOwnAddress(addr string) {
	h.ownAddress = addr
}

func (h *multiplexingHandler) handleRequest(req *http.Request) *http.Response {
	transport := &http.Transport{
		DialContext: (&net.Dialer{
			Timeout:   h.timeout,
			KeepAlive: 10 * h.timeout,
		}).DialContext,
		DisableKeepAlives:     !h.keepalive,
		TLSHandshakeTimeout:   h.timeout,
		ResponseHeaderTimeout: h.timeout,
	}
	resp, err := transport.RoundTrip(req)
	if err != nil {
		resp = &http.Response{Request: req, StatusCode: 500, Status: fmt.Sprint(err), Body: ioutil.NopCloser(strings.NewReader(fmt.Sprint(err)))}
	}
	return resp
}

func (h *multiplexingHandler) decideFinalResponse(totalCount int, successfulResponses, failedResponses []*http.Response) *http.Response {
	failedCount := len(failedResponses)
	succeededCount := len(successfulResponses)

	if failedCount == 0 && succeededCount == 0 {
		return newResponse(http.StatusServiceUnavailable, "no endpoints to query")
	}
	if succeededCount >= totalCount {
		return randomResponse(successfulResponses)
	}
	if failedCount == totalCount {
		return randomResponse(failedResponses)
	}
	if failedCount > 0 && h.allMustSucceed {
		return randomResponse(failedResponses)
	}
	return newResponse(http.StatusServiceUnavailable, "unknown error")
}

func (h *multiplexingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	ctx, cancelFunc := context.WithTimeout(context.Background(), h.timeout)
	defer cancelFunc()
	start := time.Now()
	reqId := uuid.New()
	reqLog := log.WithField("reqId", reqId)
	reqLog.Debugf("received request %v, mirroring to targets...", req.URL)
	alreadySent := false

	// Send requests to all targets in parallel and put the responses to channel
	targets := h.GetTargetAddresses()
	targetsCount := len(targets)

	responseChannel := make(chan *http.Response, targetsCount)
	wg := sync.WaitGroup{}

	for _, i := range rand.Perm(targetsCount) {
		duplicate := duplicateRequest(req).WithContext(ctx)
		if err := setRequestTarget(duplicate, targets[i], "http"); err != nil {
			reqLog.Errorf("Failed to replace new target address, error: %v", err)
			continue
		}
		wg.Add(1)
		go func() {
			responseChannel <- h.handleRequest(duplicate)
			wg.Done()
		}()
	}

	go func() {
		wg.Wait()
		close(responseChannel)
	}()

	// Check all responses from the channel
	requestCounter := 0
	var successfulResponses, failedResponses []*http.Response
mainLoop:
	for {
		select {
		case <-ctx.Done():
			if ctx.Err() != context.DeadlineExceeded {
				continue
			}
			reqLog.Error("request timed out")
			cancelFunc()
			timeoutResponse := http.Response{
				StatusCode: http.StatusGatewayTimeout,
				Body:       ioutil.NopCloser(bytes.NewBufferString("request timed out")),
			}
			sendResponse(w, &timeoutResponse)
			return
		case resp, ok := <-responseChannel:
			if !ok {
				reqLog.Debug("done processing all broadcasted requests")
				break mainLoop
			}
			requestCounter++
			if resp.StatusCode >= 400 {
				buf := new(bytes.Buffer)
				if _, err := buf.ReadFrom(resp.Body); err != nil {
					reqLog.Errorf("failed to read response body: %v", err)
				}
				errorMsg := buf.String()
				reqLog.Warnf("replica=%v request=%v status_code=%v error=%v", requestCounter, resp.Request.URL, resp.StatusCode, errorMsg)
				failedResponses = append(failedResponses, resp)
			} else {
				reqLog.Debugf("replica=%v request=%v status_code=%v", requestCounter, resp.Request.URL, resp.StatusCode)
				successfulResponses = append(successfulResponses, resp)
				if !h.allMustSucceed && !alreadySent {
					sendResponse(w, resp)
					alreadySent = true
				}
			}
		}
	}

	finalResponse := h.decideFinalResponse(requestCounter, successfulResponses, failedResponses)

	dur := time.Since(start)
	reqLog.Infof("returned final status_code=%v for request=%v with duration=%v", finalResponse.StatusCode, finalResponse.Request.URL, dur)
	requestDurationSeconds.WithLabelValues("HTTP", req.URL.Path, strconv.Itoa(finalResponse.StatusCode)).Observe(float64(dur))
	if alreadySent {
		return
	}
	sendResponse(w, finalResponse)
}


pkg/readiness/readiness.go
// Copyright 2019 FUSAKLA Martin Chodúr
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package readiness

import (
	"fmt"
	"sync"
)

func New() readiness {
	return readiness{
		ready:    fmt.Errorf("initializing"),
		readyMtx: sync.Mutex{},
	}
}

type readiness struct {
	ready    error
	readyMtx sync.Mutex
}

func (r *readiness) IsReady() error {
	r.readyMtx.Lock()
	defer r.readyMtx.Unlock()
	return r.ready
}

func (r *readiness) Ready() {
	r.readyMtx.Lock()
	defer r.readyMtx.Unlock()
	r.ready = nil
}

func (r *readiness) NotReady(err error) {
	r.readyMtx.Lock()
	defer r.readyMtx.Unlock()
	r.ready = err
}


pkg/controller/endpoints.go
// Copyright 2019 FUSAKLA Martin Chodúr
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package controller

import (
	"fmt"
	"github.com/prometheus/client_golang/prometheus"
	log "github.com/sirupsen/logrus"
	"k8s.io/apimachinery/pkg/api/meta"
	"k8s.io/apimachinery/pkg/labels"
	"k8s.io/client-go/informers"
	"k8s.io/client-go/kubernetes"
	listers "k8s.io/client-go/listers/core/v1"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/cache"
)

var (
	numberOfEndpoints = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "service_endpoint_count",
			Help: "Number of found endpoints for given service.",
		},
		[]string{"service"},
	)
)

func init() {
	prometheus.MustRegister(numberOfEndpoints)
}

func NewEndpointController(config *rest.Config, namespace *string, serviceName string, servicePortname string, updatesChannel chan *[]string) (*EndpointsController, error) {
	var informerFactory informers.SharedInformerFactory
	stopChannel := make(chan struct{})
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		return nil, err
	}
	var informerOptions []informers.SharedInformerOption
	if namespace != nil {
		informerOptions = append(informerOptions, informers.WithNamespace(*namespace))
	}
	informerFactory = informers.NewSharedInformerFactoryWithOptions(clientset, 0, informerOptions...)
	controller := EndpointsController{
		clientset: clientset,

		informer:        informerFactory.Core().V1().Endpoints().Informer(),
		lister:          informerFactory.Core().V1().Endpoints().Lister(),
		serviceName:     serviceName,
		servicePortName: servicePortname,
		updatesChannel:  updatesChannel,
		stopChannel:     stopChannel,
	}
	controller.informer.AddEventHandler(&controller)
	informerFactory.Start(stopChannel)

	return &controller, nil
}

type EndpointsController struct {
	clientset       *kubernetes.Clientset
	informer        cache.SharedIndexInformer
	lister          listers.EndpointsLister
	serviceName     string
	servicePortName string
	updatesChannel  chan *[]string
	stopChannel     chan struct{}
}

func (e *EndpointsController) ListMatchingIPs() (*[]string, error) {
	var ips []string
	endpoints, err := e.lister.List(labels.Set{}.AsSelector())
	if err != nil {
		return nil, err
	}
	for _, endpoint := range endpoints {
		if endpoint.Name != e.serviceName {
			continue
		}
		for _, subset := range endpoint.Subsets {
			var targetPort int32
			portFound := false
			for _, port := range subset.Ports {
				if port.Name == e.servicePortName {
					targetPort = port.Port
					portFound = true
				}
			}
			if !portFound {
				log.Errorf("Did not find specified port name %v in the service %v", e.servicePortName, endpoint.Name)
			}
			for _, addr := range subset.Addresses {
				ips = append(ips, fmt.Sprintf("%s:%d", addr.IP, targetPort))
			}
		}
	}
	numberOfEndpoints.WithLabelValues(e.serviceName).Set(float64(len(ips)))
	return &ips, nil
}

func (e *EndpointsController) OnAdd(obj interface{}) {
	metaObject, err := meta.Accessor(obj)
	if err != nil {
		log.Errorf("Failed to convert cache item to metaObject: %v", err)
		return
	}
	objectName := meta.AsPartialObjectMetadata(metaObject).Name
	log.Debugf("Processing endpoints update for %v", objectName)
	if e.serviceName != objectName {
		log.Debugf("Skipping non matching service: %v", objectName)
		return
	}
	ips, err := e.ListMatchingIPs()
	if err != nil {
		log.Errorf("Failed to list endpoints, error: %v", err)
		return
	}
	e.updatesChannel <- ips
}
func (e *EndpointsController) OnUpdate(_, newObj interface{}) {
	e.OnAdd(newObj)
}

func (e *EndpointsController) OnDelete(obj interface{}) {
	e.OnAdd(obj)
}

func (e *EndpointsController) StopController() {
	close(e.stopChannel)
	//close(e.updatesChannel)
}


cmd/root.go
// Copyright 2019 FUSAKLA Martin Chodúr
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cmd

import (
	"context"
	"fmt"
	"github.com/fusakla/k8s-service-broadcasting/pkg/controller"
	"github.com/fusakla/k8s-service-broadcasting/pkg/handler"
	"github.com/fusakla/k8s-service-broadcasting/pkg/readiness"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	log "github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
	"net"
	"net/http"
	"os"
	"os/signal"
	"runtime"
	"syscall"
	"time"
)

var (
	iface, metricsIface, kubeconfigPath, namespace, logLevel, serviceName, portName string
	keepalive, allMustSucceed                                                       bool
	timeout                                                                         time.Duration
	kubeconfig                                                                      *rest.Config

	rootCmd = &cobra.Command{
		Use:   "k8s-service-broadcasting",
		Short: "Broadcast HTTP to all service endpoints.",
		Long: "Tool allowing to broadcast/mirror/duplicate HTTP requests to all endpoints of Kubernetes service.\n" +
			"Waits for all of them to end and reports back failed request if any. If not returns last successful.",
		Run: runMultiplexer,
	}
)

func init() {
	cobra.OnInitialize(configure)
	rootCmd.Flags().StringVarP(&iface, "interface", "i", "0.0.0.0:8080", "Interface to listen on.")
	rootCmd.Flags().StringVarP(&metricsIface, "metrics-interface", "m", "0.0.0.0:8081", "Interface for exposing metrics.")
	rootCmd.Flags().StringVarP(&kubeconfigPath, "kubeconfig", "k", os.Getenv("KUBECONFIG"), "Location of the kubeconfig, default if in cluster config or value of KUBECONFIG env variable.")
	rootCmd.Flags().StringVarP(&serviceName, "service", "s", "", "Name of service to sed the requests to.")
	rootCmd.Flags().StringVarP(&portName, "port-name", "p", "", "Name of service port to sed the requests to.")
	rootCmd.Flags().StringVarP(&namespace, "namespace", "n", "", "Namespace to watch for.")
	rootCmd.Flags().BoolVar(&allMustSucceed, "all-must-succeed", true, "By default if any backend fails, the whole request fails. If disabled one succeeded response is enough.")
	rootCmd.Flags().StringVarP(&logLevel, "log-level", "l", "info", "Log level (debug, info, warning, ...) default info.")
	rootCmd.Flags().DurationVarP(&timeout, "timeout", "t", time.Second*10, "Timeout for mirrored requests.")
	rootCmd.Flags().BoolVar(&keepalive, "keepalive", true, "If keepalive should be enabled.")
}

// Execute executes the root command.
func Execute() error {
	return rootCmd.Execute()
}

func configure() {
	var err error
	log.SetFormatter(&log.TextFormatter{
		DisableColors: true,
		FullTimestamp: true,
	})
	log.SetOutput(os.Stdout)
	lvl, err := log.ParseLevel(logLevel)
	if err != nil {
		log.Fatalf("Failed to parse log level, error: %v", err)
	}
	log.SetLevel(lvl)

	if kubeconfigPath == "" {
		kubeconfig, err = rest.InClusterConfig()
		if err != nil {
			log.Fatalf("Failed to load kubeconfig: %v", err)
		}
	} else {
		kubeconfig, err = clientcmd.BuildConfigFromFlags("", kubeconfigPath)
		if err != nil {
			log.Fatalf("Failed to load kubeconfig: %v", err)
		}
	}
}

func runMultiplexer(cmd *cobra.Command, _ []string) {
	if err := cmd.MarkFlagRequired("service"); err != nil {
		log.Warn(err)
	}
	if err := cmd.MarkFlagRequired("port-name"); err != nil {
		log.Warn(err)
	}
	runtime.GOMAXPROCS(runtime.NumCPU())

	var status = readiness.New()

	updatesChannel := make(chan *[]string, 10)
	shutdownChannel := make(chan struct{}, 3)
	srvErrChannel := make(chan error)
	signals := make(chan os.Signal, 10)

	endpointController, err := controller.NewEndpointController(kubeconfig, &namespace, serviceName, portName, updatesChannel)
	if err != nil {
		log.Fatalf("Failed to initialize k8s endpoint watcher: %v", err)
	}

	h := handler.NewMultiplexingHandler(iface, timeout, allMustSucceed, keepalive)

	listener, err := net.Listen("tcp", iface)
	if err != nil {
		log.Fatalf("Failed to listen to %v: %v", iface, err)
	}

	server := &http.Server{
		Handler: h,
	}
	server.SetKeepAlivesEnabled(keepalive)

	go func() {
		http.Handle("/metrics", promhttp.Handler())
		http.HandleFunc("/-/healthy", func(w http.ResponseWriter, req *http.Request) { _, _ = fmt.Fprintf(w, "OK") })
		http.HandleFunc("/-/ready", func(w http.ResponseWriter, req *http.Request) {
			if status.IsReady() == nil {
				_, _ = fmt.Fprintf(w, "OK")
			} else {
				w.WriteHeader(http.StatusServiceUnavailable)
				_, _ = fmt.Fprintf(w, "NOT READY: %v", status.IsReady())
			}
		})
		if err := http.ListenAndServe(metricsIface, nil); err != nil {
			log.Errorf("Error during serving metrics, error: %v", err)
			srvErrChannel <- err
		}
	}()

	go func() {
		err := server.Serve(listener)
		if err != nil {
			if err != http.ErrServerClosed {
				log.Error(err)
				srvErrChannel <- err
			}
		}
	}()

	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)

	run := true
	for run {
		select {
		case sig, ok := <-signals:
			if !ok {
				continue
			}
			log.Infof("Received signal %v, terminating...", sig)
			shutdownChannel <- struct{}{}
		case <-srvErrChannel:
			shutdownChannel <- struct{}{}
		case <-shutdownChannel:
			status.NotReady(fmt.Errorf("shutting down"))
			ctx, cancelFunc := context.WithTimeout(context.Background(), time.Second*5)
			log.Info("Stopping k8s endpoint controller...")
			endpointController.StopController()
			log.Info("Stopping web server...")
			if err := server.Shutdown(ctx); err != nil {
				log.Errorf("Failed to gracefully stop server, error: %v", err)
			}
			ctx.Done()
			cancelFunc()
			run = false
			break
		case ips, ok := <-updatesChannel:
			if ok {
				status.Ready()
				log.Infof("Updating targets with new addresses: %v", *ips)
				h.SetTargetAddresses(*ips)
				continue
			}
		}
	}
}
